### PR TITLE
Fixed an edge case that could result in negative GQs when in USE_POSTERIOR_PROBABILITIES mode

### PR DESCRIPTION
**Pull request by jamesemery**
Resolves one of the issues encountered when merging the DRAGEN-GATK gvcfs the #7108.

### CODE REVIEW

**Comment by droazen** [comment]
@ldgauthier Would you like to review this one as well, since you reviewed https://github.com/broadinstitute/gatk/pull/7148?  It's just some added `Math.min()` statements by the look of it...

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 1]
```diff
@@ -340,22 +341,23 @@ public static void makeGenotypeCall(final int ploidy,
         }
     }
 
-    private static double getGQLog10FromPosteriors(final int bestGenotypeIndex, final double[] /**/log10Posteriors) {
+    @VisibleForTesting
+    static double getGQLog10FromPosteriors(final int bestGenotypeIndex, final double[] /**/log10Posteriors) {
```
Comment (ldgauthier) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:345]: @vruano James and I are both wondering why we're allowing a `bestGenotypeIndex` that isn't the array max.

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java [hunk 2]
```diff
@@ -340,22 +341,23 @@ public static void makeGenotypeCall(final int ploidy,
...
                 case 1: return CommonInfo.NO_LOG10_PERROR;
                 case 2: return bestGenotypeIndex == 0 ? log10Posteriors[1] : log10Posteriors[0];
-                case 3: return Math.min(0, MathUtils.log10SumLog10(
+                case 3: return Math.min(0.0, MathUtils.log10SumLog10(
                                  log10Posteriors[ bestGenotypeIndex == 0 ? 2 : bestGenotypeIndex - 1],
                                  log10Posteriors[ bestGenotypeIndex == 2 ? 0 : bestGenotypeIndex + 1]));
                 default:
                     if (bestGenotypeIndex == 0) {
-                        return MathUtils.log10SumLog10(log10Posteriors, 1, log10Posteriors.length);
+                        return Math.min(0.0, MathUtils.log10SumLog10(log10Posteriors, 1, log10Posteriors.length));
```
Comment (ldgauthier) [inline src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:358]: Another issue for @vruano : If we have to throw a min on this, then the normalization isn't happening in the right place.  According to my calculations (and as seen in James's unit test) the common case where this comes into play is where there are more than two equally likely genotypes, so the log10 probabilities get normalized to [0, 0, 0, -something, more negative somethings], the best is a zero and then there are two not bests that are zero and the log10SumLog10(0,0) makes a log10(2) ~ 0.3 > 0 which leads to a negative GQ in Phred-space.  Elsewhere we use the MathUtils.secondSmallestMinusSmallest, which doesn't have this problem.  I don't think the log10SumLog10 is the right thing to do.

**Review by ldgauthier — CHANGES_REQUESTED** [review]

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 1 above)
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:345]: I think this is just a optimization..... looking where this method is used, is often the case that that max index has already been calculated earlier on the very same probs/lks array for another reason and by taken it as an argument we are saving ourselves the recalculation cost. 

Have you seen anywhere where this rule is broken (i.e. the index was calculated from a different array or the array was altered after the index was calculated?)

I can see that it is a code smell and so If you prefer to recalculate i'm ok with that. Perhaps we should consider to encapsulate prob/lk array in an object that would keep track of best and second best but that perhaps is over-kill for now.

Comment (ldgauthier) [reply src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:345]: I haven't seen any inconsistent cases, but can we at least specify in some javadoc that the index should agree with the minimum of the posterior array being passed in?

File: src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java (hunk 2 above)
Comment (vruano) [reply src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:358]: Based on the definition of GQ in the VCF we are supposed to do that.

Comment (ldgauthier) [reply src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVariantContextUtils.java:358]: Given that we output integer GQ values (as in the spec), the "smallest minus second smallest" will be the same as 1-Prob(GT)/Z unless the PLs are [0,0,0] right?

**Comment by jamesemery** [comment]
@ldgauthier @vruano Given that this is still fixing a rather glaring bug in the current behavior and we should probably get a fix of some ilk in before the next release of GATK I'm going to push the decision to rework the GT calculation to an open issue ticket in favor of getting this branch merged.

**Comment by ldgauthier** [comment]
I have merged many a PR by opening new issues, so I would be a hypocrite to object.  Sounds good to me @jamesemery.

**Comment by droazen** [comment]
@jamesemery What's the status of this one as well? Close to a merge?

**Comment by jamesemery** [comment]
@ldgauthier @vruano can i get a passing review on this so it can be merged? the failing tests were transient errors.

**Review by vruano — COMMENTED** [review]
No objections

**Review by vruano — APPROVED** [review]

**Pull request merged**